### PR TITLE
crimson/sesatore: simplify epm related transaction structures, fix some bugs

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

File: src/crimson/os/seastore/transaction.h [hunk 1]
```diff
@@ -68,16 +63,7 @@ class Transaction {
...
-      // We decide not to remove it from fresh_block_list because touching this
-      // will affect relative paddrs, and it should be rare to retire a fresh
-      // extent.
       ref->state = CachedExtent::extent_state_t::INVALID;
-      if (ref->is_inline()) {
-	write_set.erase(*ref);
-      } else {
-	// if ref is not relative, it must be in the delayed set
-	delayed_set.erase(*ref);
-      }
```
Comment (xxhdx1985126) [inline src/crimson/os/seastore/transaction.h:80]: I'm not quite following, don't we have to remove the extent from `Transaction::write_set` when it's retired at the state CachedExtent::extent_state_t::INITIAL_WRITE_PENDING? I thought extents in this state can be in the write_set.

File: src/crimson/os/seastore/transaction.h [hunk 2]
```diff
@@ -228,9 +234,9 @@ class Transaction {
     delayed_temp_offset = 0;
     read_set.clear();
     invalidate_clear_write_set();
-    fresh_block_list.clear();
     mutated_block_list.clear();
     delayed_alloc_list.clear();
+    inline_block_list.clear();
```
Comment (xxhdx1985126) [inline src/crimson/os/seastore/transaction.h:254]: I think ool_block_list also have to be cleared, am I right?

File: src/crimson/os/seastore/scanner.cc [hunk 3]
```diff
@@ -116,7 +116,7 @@ Scanner::scan_valid_records_ret Scanner::scan_valid_records(
     cursor.offset.offset = segment_manager.get_block_size();
   }
   auto retref = std::make_unique<size_t>(0);
-  auto budget_used = *retref;
+  auto &budget_used = *retref;
```
Comment (xxhdx1985126) [inline src/crimson/os/seastore/scanner.cc:119]: I think this should fix https://tracker.ceph.com/issues/52582, am I right?

File: src/crimson/os/seastore/transaction.h (hunk 1 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction.h:80]: I don't see a problem with leaving it in the write_set, and it simplifies the rules.

File: src/crimson/os/seastore/scanner.cc (hunk 3 above)
Comment (athanatos) [reply src/crimson/os/seastore/scanner.cc:119]: Yep, I'll add that to the commit message.

File: src/crimson/os/seastore/transaction.h (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/transaction.h:254]: Yep, fixed.

**Review by xxhdx1985126 — APPROVED** [review]

**Review by cyx1231st — APPROVED** [review]

**Pull request merged**